docs/src/transactions/transaction-kernel.md

```diff
@@ -0,0 +1,104 @@
+# The Transaction Kernel Program
+The transaction kernel program is responsible to execute a Miden rollup transaction within the Miden VM. Therefore, it is written in [MASM](https://0xpolygonmiden.github.io/miden-vm/user_docs/assembly/main.html) and it is defined as MASM [kernel](https://0xpolygonmiden.github.io/miden-vm/user_docs/assembly/execution_contexts.html#kernels). The kernel provides context-sensitive security preventing unwanted read and write access. It defines a set of procedures which can be invoked from other [contexts](https://0xpolygonmiden.github.io/miden-vm/user_docs/assembly/execution_contexts.html#execution-contexts) - e.g., notes - to be executed in the root context.  
+
+In general, the kernel's procedures must reflect everything users might want to do in executing transactions, from transferring assets to complex smart contract interactions with custom code. Learn more about available procedures and contexts [here](transaction-procedures.md).
+
+The kernel has a well-defined structure which must do the following:
+
+1. **Prologue**: prepares the transaction for processing by parsing the transaction data and setting up the root context.
+2. **Note Processing**: executes the note processing loop which consumes each InputNote and invokes the note script of each note.
+3. **Transaction Script Processing**: executes the optional transaction script.
+4. **Epilogue**: finalizes the transaction by computing the created notes commitment, the final account hash, asserting asset invariant conditions and asserting the nonce rules are upheld.
+
+<p align="center">
+    <img src="../diagrams/architecture/transaction/Transaction_program.png" style="width: 75%;">
+</p>
+
+## Memory Layout 
+Miden Assembly doesn't provide dynamic memory allocation. In total, a MASM program has `2^{32}` memory slots. Therefore, we need to pre-allocate areas in which we store all necessary information. The Miden kernel uses memory offsets to manage the allocation. The root context, in other words the kernel, has access to all of those memory slots. 
+
+<p align="center">
+    <img src="../diagrams/architecture/transaction/Memory_layout_kernel.png" style="width: 75%;">
+</p>
+
+The book keeping section is needed to keep track of variables which are used internally by the transaction kernel. Global inputs are the block hash and the global nullifier commitment. The rest will be explained in the next section.
+
+## The Prologue
+The transaction prologue is executed at the beginning of a transaction. It needs to accomplish the following tasks:
```

Comment:
One important thing to mention (probably before we get into prologue) is what are the public inputs/outputs of the tx kernel (i.e., stack inputs into the prologue and stack outputs from the epilogue).

Reply:
added an input and output section. Might be too much detail, let's see.